src/cancellation_token.c

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/logger.h"
+
+#include "c_pal/thandle.h"
+#include "c_pal/gballoc_hl.h"
+#include "c_pal/gballoc_hl_redirect.h"
+#include "c_pal/interlocked.h"
+
+#include "c_util/tcall_dispatcher_cancellation_token_cancel_call.h"
+#include "c_util/cancellation_token.h"
+
+#define CANCELLATION_TOKEN_STATE_VALUES \
+    CANCELLATION_TOKEN_STATE_NOT_CANCELED, \
+    CANCELLATION_TOKEN_STATE_CANCELED
+
+MU_DEFINE_ENUM(CANCELLATION_TOKEN_STATE, CANCELLATION_TOKEN_STATE_VALUES);
+MU_DEFINE_ENUM_STRINGS(CANCELLATION_TOKEN_STATE, CANCELLATION_TOKEN_STATE_VALUES);
+
+typedef struct CANCELLATION_TOKEN_TAG
+{
+    volatile_atomic int32_t state;
```

Comment:
>volatile_atomic int32_t state; [](http://example.com/codeflow?start=4&length=30)

consider putting this state in a union with the `CANCELLATION_TOKEN_STATE` enum, just because it is easier on the debug eyes.

Reply:
Took a shot at this. Hope this is what you meant!

```c
typedef union CANCELLATION_TOKEN_STATE_VALUE_TAG
{
    volatile_atomic int32_t state;
    volatile_atomic CANCELLATION_TOKEN_STATE state_enum;
} CANCELLATION_TOKEN_STATE_VALUE;

typedef struct CANCELLATION_TOKEN_TAG
{
    CANCELLATION_TOKEN_STATE_VALUE state;
    TCALL_DISPATCHER(CANCELLATION_TOKEN_CANCEL_CALL) cancel_notification_call_dispatcher;
} CANCELLATION_TOKEN;
```